src/OrchardCore/OrchardCore.Users.Core/Services/Rfc6238AuthenticationService.cs

```diff
@@ -0,0 +1,154 @@
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.Net;
+using System.Security.Cryptography;
+using System.Text;
+using OrchardCore.Modules;
+using OrchardCore.Users.Models;
+
+namespace OrchardCore.Users.Services;
+
+/// <summary>
+/// The following code is influenced by <see href="https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Extensions.Core/src/Rfc6238AuthenticationService.cs"/>
+/// </summary>
+public sealed class Rfc6238AuthenticationService
```

Comment:
@MikeAlhayek I'm not sure if this the right direction to implement own `Rfc6238AuthenticationService` version. Original `Rfc6238AuthenticationService` service do not allow changing time span and pin length by design. **It is for security reason.**

Please imagine the scenario when somebody, will use 2 digits code and time span set to let's say to 10 minutes... You would be able to guess the code by trying multiple times all possible codes.

Other than that, I've extracted your service and adjusted it a bit (to be able to pass UtcNow) to test a few possible edge cases, and I found that something is wrong. I used 10 minutes time span, and after this time, I'm getting that my PIN is still valid:

your service tested:
![image](https://github.com/OrchardCMS/OrchardCore/assets/6403130/63d4d775-ead3-4324-9d98-d768d43584bf)

original service tested:
![image](https://github.com/OrchardCMS/OrchardCore/assets/6403130/92534e82-35d7-478e-a5b1-ec2e8e238f86)

I will try to find some time to read your code and check it deeper. But as I said, I'm not sure if this is the right way. We should use providers which allow using configurable timespans. Those providers are not good for all scenarios, because codes produced are much longer (so they are not suitable for SMS). It is because `creation time` is a part of encrypted (protected) code. You can read more [here](https://github.com/dotnet/aspnetcore/blob/79ef5e329b1e31c3775a1977798253cc8f7da6cc/src/Identity/Core/src/DataProtectorTokenProvider.cs#L116).

Reply:
@lampersky I removed support for pins length < 6 characters. The Algorithm that generates pins validates with pics generated outside the expiration date-time because the clock that generates tokens could be different from the clock that validates the token. More about [RFC6238 ](https://datatracker.ietf.org/doc/html/rfc6238)

Reply:
Any way, I removed it and cleanup some other stuff.